File: js/geometria/action/GEraseSelectionAction.js
/**
 * Copyright (C) 2000-2014 Geometria Contributors
 * http://geocentral.net/geometria
 * 
 * Geometria is free software released under the MIT License
 * http://opensource.org/licenses/MIT
 */
define([
    "dojo/_base/lang",
    "geometria/GDictionary",
    "geometria/GFace",
    "geometria/GFiguresContainer",
    "geometria/GLine",
    "geometria/GMainContainer",
    "geometria/GNotepadContainer",
    "geometria/GPoint",
    "geometria/GUtils"
], function(lang, dict, GFace, figuresContainer, GLine, mainContainer, notepadContainer, GPoint,
            utils) {

    var validate = function(props) {
        var solid = figuresContainer.getFigure(props.figureName).solid;
        var removedLines = {};
        var valid = true;
        $.each(props.lines, function() {
            stickCodes = this;
            var faces = solid.facesThroughPoints(stickCodes);
            if (utils.propCount(faces) > 1) {
                valid = false;
                return false;
            }
            var face = utils.anyProp(faces);
            var faceCode = face.getFaceCode();
            var removedLs = removedLines[faceCode];
            if (!removedLs) {
                removedLs = [];
                removedLines[faceCode] = removedLs;
            }
            var line = face.lineThroughPoints(stickCodes[0], stickCodes[1]);
            removedLs.push(line.labels.slice(0));
        });
        if (!valid) {
            return null;
        }
        var removedPoints = {};
        $.each(props.points, function(index, label) {
            var p = solid.points[label];
            removedPoints[label] = { crds: p.crds, affectedLines: [] };
        });
        $.each(props.points, function(index, label) {
            var faces = solid.facesThroughPoints([label]);
            $.each(faces, function() {
                var face = this;
                var lines = face.linesThroughPoint(label);
                var removedLs = removedLines[face.getFaceCode()];
                if (removedLs) {
                    $.each(removedLs, function() {
                        var labels = this;
                        var line = face.lineThroughPoints(labels[0], labels[1]);
                        utils.remove(lines, line);
                    });
                }
                $.each(lines, function() {
                    var line = this;
                    if (line.firstLabel() == label || line.lastLabel() == label) {
                        valid = false;
                        return false;
                    }
                });
                if (!valid) {
                    return false;
                }
            });
            if (!valid) {
                return false;
            }
        });
        if (valid) {
            var outProps = lang.mixin({
                removedPoints: removedPoints,
                removedLines: removedLines
            }, props);
            return outProps;
        }
        return null;
    };

    var validateExternal = function(props) {
        var figure = figuresContainer.getFigure(props.figureName);
        if (!figure) {
            return null;
        }
        if (!Array.isArray(props.points) || !Array.isArray(props.lines) || !props.figureName) {
            return null;
        }
        var valid = true;
        $.each(props.lines, function() {
            if (!Array.isArray(this) || this.length != 2) {
                valid = false;
                return false;
            }
        });
        if (valid) {
            return validate(props);
        }
        return null;
    };
    
    var validateSelection = function() {
        var figure = figuresContainer.getSelectedFigure();
        var solid = figure.solid;
        var selectedElements = [];
        var pointLabels = [];
        var stickCodes = [];
        $.each(solid.selection, function(code, element) {
            if (element instanceof GPoint) {
                pointLabels.push(element.label);
            }
            else if (!(element instanceof GFace)) {
                stickCodes.push([element.p1.label, element.p2.label]);
            }
        });
        if (pointLabels.length > 0 || stickCodes.length > 0) {
            var props = {
                figureName: figure.name,
                points: pointLabels,
                lines: stickCodes
            }
            return validate(props);
        }
        return null;
    };

    var apply = function(props) {
        var figure = figuresContainer.getFigure(props.figureName);
        var solid = figure.solid;
        var danglingPoints = {};
        $.each(props.removedLines, function(faceCode, removedLs) {
            var face = solid.getFace(faceCode);
            $.each(removedLs, function() {
                var labels = this;
                var line = face.lineThroughPoints(labels[0], labels[1]);
                var danglingLabels = [];
                face.removeLine(line, danglingLabels);
                $.each(danglingLabels, function(index, label) {
                    danglingPoints[label] = solid.points[label];
                });
            });
        });
        $.each(props.removedPoints, function(label, removedP) {
            var affectedLs = [];
            var faces = solid.facesThroughPoints([label]);
            $.each(faces, function() {
                var face = this;
                var lines = face.linesThroughPoint(label);
                if (lines.length > 0) {
                    var removedLs = props.removedLines[face.getFaceCode()];
                    if (removedLs) {
                        $.each(removedLs, function() {
                            var labels = this;
                            var line = face.lineThroughPoints(labels[0], labels[1]);
                            utils.remove(lines, line);
                        });
                    }
                    affectedLs = affectedLs.concat(lines);
                }
            });
            if (affectedLs.length > 0) {
                $.each(affectedLs, function() {
                    var line = this;
                    utils.remove(line.labels, label);
                    removedP.affectedLines.push(line.toJson());
                });
            }
            delete solid.points[label];
        });
        $.each(danglingPoints, function(label) {
            if (!props.removedPoints[label]) {
                var crds = solid.points[label].crds;
                props.removedPoints[label] = { crds: crds, affectedLines: [] };
                delete solid.points[label];
            }
        });
        $.each(props.removedPoints, function(label) {
            notepadContainer.pointRemoved(label, props.figureName, props.executeId);
        });
        solid.makeConfig();
        solid.selection = {};
        figure.draw(true);
        return props;
    };

    return {

        loggable: true,
    
        figureSpecific: true,

        icon: "geometriaIcon24 geometriaIcon24Erase",

        label: dict.get("action.EraseSelection"),
        
        execute: function() {
            var figure = figuresContainer.getSelectedFigure();
            var solid = figure.solid;
            var selectionProps = validateSelection();
            selectionProps.executeId = this.getExecuteId();
            var outProps = apply(selectionProps);
            notepadContainer.update();
            mainContainer.setDocumentModified(true);
            return outProps;
        },

        validateSelection: validateSelection,
        
        undo: function(props) {
            var figure = figuresContainer.getFigure(props.figureName);
            var solid = figure.solid;
            solid.selection = {};
            $.each(props.removedPoints, function(label, removedP) {
                solid.points[label] = new GPoint(label, removedP.crds);
            });
            $.each(props.removedLines, function(faceCode, removedLs) {
                var face = solid.getFace(faceCode);
                $.each(removedLs, function() {
                    var labels = this;
                    var line = new GLine(labels);
                    face.lines.push(line);
                });
            });
            $.each(props.removedPoints, function(label, removedP) {
                if (removedP.affectedLines) {
                    $.each(removedP.affectedLines, function() {
                        var labels = this;
                        var lines = solid.linesThroughPoints(labels[0], labels[1]);
                        var p = solid.points[label];
                        $.each(lines, function() {
                            this.insert(p, solid);
                        });
                    });
                }
            });
            $.each(props.removedPoints, function(label) {
                notepadContainer.removePointUndone(props.executeId);
            });
            notepadContainer.update();
            solid.makeConfig();
            figure.draw(true);
            figuresContainer.select(props.figureName);
        },
        
        playBack: function(props, external) {
            var outProps = lang.mixin({}, props);
            if (external) {
                outProps.executeId = this.base.getExecuteId();
                outProps = validateExternal(outProps);
                if (!outProps) {
                    return null;
                }
            }
            outProps = apply(outProps);
            if (!external) {
                notepadContainer.update();
            }
            return outProps;
        },

        toTooltip: function(props) {
            return this.toLog(props);
        },
        
        toLog: function(props) {
            return dict.get("EraseSelectionInFigure", props.figureName);
        },

        updateState: function() {
            this.base.enabled = figuresContainer.getSelectedFigure() &&
                this.validateSelection() && true;
        },

        toJson: function(props) {
            return {
                "action": "eraseSelectionAction",
                "props": {
                    "figureName": props.figureName,
                    "points": props.points,
                    "lines": props.lines
                }
            };
        }
    };
});

File: archive/3.0/src/net/geocentral/geometria/action/GEraseSelectionAction.java
/**
 * Copyright 2000-2010 Geometria Contributors
 * http://geocentral.net/geometria
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License
 * http://www.gnu.org/licenses
 */
package net.geocentral.geometria.action;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import net.geocentral.geometria.model.GDocument;
import net.geocentral.geometria.model.GFace;
import net.geocentral.geometria.model.GFigure;
import net.geocentral.geometria.model.GLine;
import net.geocentral.geometria.model.GPoint3d;
import net.geocentral.geometria.model.GSelectable;
import net.geocentral.geometria.model.GSolid;
import net.geocentral.geometria.model.GStick;
import net.geocentral.geometria.util.GDictionary;
import net.geocentral.geometria.util.GGraphicsFactory;

import org.apache.log4j.Logger;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

public class GEraseSelectionAction implements GLoggable, GFigureAction {

    private String figureName;

    private List<String> pointLabels;

    private List<StringPair> stickLabels;

    private GDocument document;

    private GFigure figure;

    private GSolid solid;

    private Map<GFace, List<GLine>> removedLines;

    private Map<GPoint3d, List<GLine>> removedPoints;

    private static Logger logger = Logger.getLogger("net.geocentral.geometria");

    public boolean execute(GDocumentHandler documentHandler,
            boolean quietMode) {
        logger.info("");
        document = documentHandler.getActiveDocument();
        if (quietMode) {
            try {
                validateApply();
            }
            catch (Exception exception) {
                return false;
            }
        }
        else {
            GFigure figure = document.getSelectedFigure();
            figureName = figure.getName();
            GSolid solid = figure.getSolid();
            Set<GSelectable> selection = solid.getSelection();
            if (selection.isEmpty())
                return false;
            pointLabels = new ArrayList<String>();
            stickLabels = new ArrayList<StringPair>();
            for (GSelectable element : selection) {
                if (element instanceof GStick) {
                    StringPair sp = new StringPair(((GStick)element).label1,
                            ((GStick)element).label2);
                    stickLabels.add(sp);
                }
                else if (element instanceof GPoint3d) {
                    pointLabels.add(((GPoint3d)element).getLabel());
                }
            }
            try {
                validateApply();
            }
            catch (Exception exception) {
                GGraphicsFactory.getInstance().showErrorDialog(exception.getMessage());
                return false;
            }
            solid.clearSelection();
        }
        document.getNotepad().selectionErased(this);
        documentHandler.notepadChanged();
        document.setSelectedFigure(figureName);
        document.getSelectedFigure().repaint();
        if (!quietMode)
            documentHandler.setDocumentModified(true);
        logger.info(figureName + ", " + pointLabels + ", " + stickLabels);
        return true;
    }

    public void validateApply() throws Exception {
        logger.info("");
        figure = document.getFigure(figureName);
        solid = figure.getSolid();
        removedLines = new LinkedHashMap<GFace, List<GLine>>();
        for (StringPair sp : stickLabels) {
            Collection<GFace> faces = solid.facesThroughPoints(new String[] {
                    sp.s1, sp.s2 });
            if (faces.size() > 1) {
                logger.info("Edge: " + sp);
                throw new Exception(GDictionary.get("CannotEraseEdgeInFigure",
                        sp.s1 + sp.s2, figureName));
            }
            GFace face = faces.iterator().next();
            List<GLine> removedLs = removedLines.get(face);
            if (removedLines.get(face) == null) {
                removedLs = new ArrayList<GLine>();
                removedLines.put(face, removedLs);
            }
            GLine line = face.lineThroughPoints(sp.s1, sp.s2);
            removedLs.add(line);
        }
        removedPoints = new LinkedHashMap<GPoint3d, List<GLine>>();
        for (String s : pointLabels) {
            GPoint3d p = solid.getPoint(s);
            removedPoints.put(p, null);
        }
        for (String label : pointLabels) {
            Collection<GFace> faces = solid.facesThroughPoint(label);
            for (GFace face : faces) {
                List<GLine> lines = face.linesThroughPoint(label);
                List<GLine> removedLs = removedLines.get(face);
                if (removedLs != null)
                    lines.removeAll(removedLs);
                for (GLine line : lines) {
                    if (line.firstLabel().equals(label)
                            || line.lastLabel().equals(label)) {
                        logger.info("End of line: " + label);
                        throw new Exception(GDictionary.get(
                            "CannotEraseEndOfLineInFigure", label, figureName));
                    }
                }
            }
        }
        // Apply
        Set<GPoint3d> danglingPoints = new LinkedHashSet<GPoint3d>();
        for (GFace face : removedLines.keySet()) {
            List<GLine> removedLs = removedLines.get(face);
            for (GLine line : removedLs) {
                List<String> danglingLabels = new ArrayList<String>();
                face.removeLine(line, danglingLabels);
                for (String label : danglingLabels)
                    danglingPoints.add(solid.getPoint(label));
            }
        }
        for (GPoint3d p : removedPoints.keySet()) {
            List<GLine> affectedLines = new ArrayList<GLine>();
            Collection<GFace> faces = solid.facesThroughPoint(p.getLabel());
            for (GFace face : faces) {
                List<GLine> lines = face.linesThroughPoint(p.getLabel());
                if (lines.isEmpty())
                    continue;
                List<GLine> removedLs = removedLines.get(face);
                if (removedLs != null)
                    lines.removeAll(removedLs);
                affectedLines.addAll(lines);
            }
            if (!affectedLines.isEmpty()) {
                for (GLine line : affectedLines)
                    line.remove(p.getLabel());
                removedPoints.put(p, affectedLines);
            }
            solid.removePoint(p.getLabel());
        }
        for (GPoint3d p : danglingPoints) {
            solid.removePoint(p.getLabel());
            removedPoints.put(p, null);
        }
        solid.makeConfig();
    }

    public void undo(GDocumentHandler documentHandler) {
        logger.info("");
        solid.clearSelection();
        for (GPoint3d p : removedPoints.keySet())
            solid.undoRemovePoint(p);
        for (GFace face : removedLines.keySet()) {
            List<GLine> removedLs = removedLines.get(face);
            for (GLine line : removedLs)
                face.addLine(line);
        }
        for (GPoint3d p : removedPoints.keySet()) {
            List<GLine> affectedLines = removedPoints.get(p);
            if (affectedLines != null) {
                for (GLine line : affectedLines)
                    line.insert(p, solid);
            }
        }
        document.getNotepad().eraseSelectionUndone(this);
        documentHandler.notepadChanged();
        solid.makeConfig();
        figure.repaint();
        document.setSelectedFigure(figureName);
        logger.info(figureName + ", " + pointLabels + ", " + stickLabels);
    }

    public GLoggable clone() {
        GEraseSelectionAction action = new GEraseSelectionAction();
        action.figureName = figureName;
        action.pointLabels = new ArrayList<String>();
        for (String label : pointLabels)
            action.pointLabels.add(label);
        action.stickLabels = new ArrayList<StringPair>();
        for (StringPair sp : stickLabels)
            action.stickLabels.add(new StringPair(sp.s1, sp.s2));
        return action;
    }

    public String toLogString() {
        return GDictionary.get("EraseSelectedElementsInFigure", figureName);
    }

    public void make(Element node) throws Exception {
        logger.info("");
        NodeList ns = node.getElementsByTagName("figureName");
        if (ns.getLength() == 0) {
            logger.error("No figure name");
            throw new Exception();
        }
        figureName = ns.item(0).getTextContent();
        ns = node.getElementsByTagName("pLabel");
        pointLabels = new ArrayList<String>();
        for (int i = 0; i < ns.getLength(); i++) {
            String label = ns.item(i).getTextContent();
            pointLabels.add(label);
        }
        NodeList nns = node.getElementsByTagName("sLabels");
        stickLabels = new ArrayList<StringPair>();
        for (int i = 0; i < nns.getLength(); i++) {
            String value = nns.item(i).getTextContent();
            String[] tokens = value.split(" ");
             if (tokens.length != 2) {
                 logger.error(Arrays.asList(tokens));
                 throw new Exception();
             }
            StringPair sp = new StringPair(tokens[0], tokens[1]);
            stickLabels.add(sp);
        }
    }

    public void serialize(StringBuffer buf) {
        logger.info("");
        buf.append("\n<action>");
        buf.append("\n<className>");
        buf.append(this.getClass().getSimpleName());
        buf.append("</className>");
        buf.append("\n<figureName>");
        buf.append(figureName);
        buf.append("</figureName>");
        for (String label : pointLabels) {
            buf.append("\n<pLabel>");
            buf.append(label);
            buf.append("</pLabel>");
        }
        for (StringPair sp : stickLabels) {
            buf.append("\n<sLabels>");
            buf.append(sp.s1);
            buf.append(" ");
            buf.append(sp.s2);
            buf.append("</sLabels>");
        }
        buf.append("\n</action>");
    }

    public String getShortDescription() {
        return GDictionary.get("eraseSelection");
    }

    public String getFigureName() {
        return figureName;
    }

    public Collection<GPoint3d> getRemovedPoints() {
        return removedPoints.keySet();
    }

    class StringPair {

        public String s1;

        public String s2;

        public StringPair(String s1, String s2) {
            this.s1 = s1;
            this.s2 = s2;
        }
        
        public String toString() {
            return s1 + ", " + s2;
        }
    }
}
